core/unix/memquery_test.h

```diff
@@ -0,0 +1,344 @@
+        fake_memquery_result *cur = &cur_bounds->iters[i];
+        memquery_iter_t iter = cur->iter_result;
+        if (iter.vm_start <= base && base < iter.vm_end) {
+            if (cur->is_header) {
+                *out_base = cur->mod_base;
+                *out_max_end = cur->mod_end;
+                return true;
+            }
+        }
+    }
+    ASSERT(false && "UNKNOWN BASE PC");
+    return false;
+}
+
+static const memquery_library_bounds_funcs fake_memquery_library_bounds_funcs = {
+    .memquery_iterator_start = memquery_iterator_start_test,
+    .memquery_iterator_next = memquery_iterator_next_test,
+    .memquery_iterator_stop = memquery_iterator_stop_test,
+    .module_is_header = module_is_header_test,
+    .module_walk_program_headers = module_walk_program_headers_test,
+};
+
+static bool
+run_single_memquery_test(const memquery_library_bounds_test *test)
+{
+    cur_bounds = test;
+
+    char fulldir[MAXIMUM_PATH];
+    char filename[MAXIMUM_PATH];
+
+    app_pc start = test->in_start;
+    app_pc end;
+
+    int got_return = memquery_library_bounds_by_iterator_internal(
+        test->in_name, &start, &end, test->want_fulldir ? fulldir : NULL,
+        BUFFER_SIZE_ELEMENTS(fulldir), test->want_filename ? filename : NULL,
+        BUFFER_SIZE_ELEMENTS(filename), &fake_memquery_library_bounds_funcs);
+
+    EXPECT(start, test->want_start);
+    EXPECT(end, test->want_end);
+    EXPECT(got_return, test->want_return);
+    if (test->want_fulldir) {
+        EXPECT_STR(fulldir, test->want_fulldir, BUFFER_SIZE_ELEMENTS(fulldir));
+    }
+    if (test->want_filename) {
+        EXPECT_STR(filename, test->want_filename, BUFFER_SIZE_ELEMENTS(filename));
+    }
+
+    cur_bounds = NULL;
+    return true;
+}
+
+#endif // STANDALONE_UNIT_TEST
+
+#ifdef RECORD_MEMQUERY
+
+#    ifndef RECORD_MEMQUERY_RESULTS_FILE
+#        define RECORD_MEMQUERY_RESULTS_FILE "/tmp/memquery_results.txt"
+#        define RECORD_MEMQUERY_TESTS_FILE "/tmp/memquery_tests.txt"
+#    endif
+
+// We rename memquery_library_bounds_by_iterator to
+// real_memquery_library_bounds_by_iterator in a define below.
+#    define record_memquery_library_bounds_by_iterator memquery_library_bounds_by_iterator
+
+int
+real_memquery_library_bounds_by_iterator(const char *name, app_pc *start /*IN/OUT*/,
+                                         app_pc *end /*OUT*/,
+                                         char *fulldir /*OPTIONAL OUT*/,
+                                         size_t fulldir_size,
+                                         char *filename /*OPTIONAL OUT*/,
+                                         size_t filename_size);
+
+/* record_memquery_library_bounds_by_iterator wraps
+ * memquery_library_bounds_by_iterator, printing out test case data for the
+ * memquery unit test directly to STDERR. See memquery_test_data.h for a
+ * selection of tests.
+ *
+ * To use this function to generate more test cases, compile with RECORD_MEQUERY
```

Comment:
We'd also want to build this as part of the regression tests to avoid bitrot (best not to have a config we want to maintain that is never automatically tested).  To avoid slowing down Travis it could just be a CDash build: add it to runsuite.cmake inside `if (DO_ALL_BUILDS)` (see `release-internal-32` e.g.).

Reply:
+1, thanks for the good suggestion.